Documentation/proposal/fetchers_refactor.md

```diff
@@ -0,0 +1,125 @@
+
+
+
+## Distribution Handlers plugins
+
+With this approach the distributions handler will live in the stage0. If there's the need to make them plugins the main tricky points is that they need to talk with an image store, so it'll be difficult to code them in a different language than go.
+
+## Transport handler and transport plugins
+
+Instead transport logic should be easily coded in a plugin if a good plugin interface will be defined.
+
+### Implementation
+
+The idea (also to handle transport caching and other things) is to implement the transport plugin logic in this way:
+
+* Transport handler: will live in the stage0 and be the entrypoint for the distribution handlers. It'll be just one and transport independent. Its purpose is the prepare the input data to send to a transport plugin, call it and handle plugin outputs.
+* Transport plugin: will read the input data provided by the transport handler, use it if needed, return the output data.
+
+### Transport level caching
+
+For future store work I'd like to split out all the information not related to an image.
+
+Now the remote table in the current store serves two purposes:
+ * A sort of reference store (url -> digest)
+ * transport caching data (now only http etag and Cache Control max-age)
+
+The first one will be covered with the ref store using the distribution concept in #2953 
+
+For the latter I'd like to create a caching transport store facility that transports implementations should use the fetch/save caching data.
+Transport caching can be used to fetch objects that given the same transport address may change (like ACIs). To avoid fetching them every time we can use transport provided caching information.
+OCI blobs, being referenced by their digest can be considered immutable and so no caching information will be required.
+ 
+The idea (that I'm going to try also in the image store) will be to not use a relational db like ql but a k/v store like boltdb (https://github.com/boltdb/bolt)  (I'll open a new issue to talk about pro/cons of this solution)
+
+For all the transport types the key will be the transport url (http://... file:// etc...) it will be associated with a blob digest and a json documents that will contain the caching information.
+
+**NOTE**
+Since the json document can contain arbitrary data, is there the need to standardize it for every transport type so it can be used as input and provided as output by all the transport plugin for that specific transport type?
+In this way multiple independent plugin implementations for the same transport will use the same data, if this isn't standardized and every transport plugin uses it's own kind of data since (it will be saved in the transport caching store) if a transport plugin is swapped with another one using diffrent data, this data won't be recognised by that transport plugin causing possible problems.
```

Comment:
I agree, the possibility to receive not understandable data by a plugin should just be stated somewhere so plugin writers know that they should handle this case.